proposals/0000-exception-backtraces.rst

```diff
@@ -0,0 +1,255 @@
+
+We will consider these two pieces in turn.
+
+Extending ``SomeException``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+All exceptions in Haskell are represented by the
+``Control.Exception.SomeException`` type. Consequently, this serves as a
+natural place to attach backtrace information.  We propose to extend
+``SomeException`` as follows: ::
+
+    data SomeException where
+        SomeExceptionWithLocation
+          :: forall e. Exception e
+          => Maybe Backtrace   -- ^ backtrace, if available
+          -> e                 -- ^ the exception
+          -> SomeException
+
+We will leave the details of the representation of the ``Backtrace`` type for
+later.
+
+To retain backwards compatibility, we redefine the old ``SomeException`` data
+constructor as a pattern synonym: ::
+
+    pattern SomeException e <- SomeExceptionWithLocation _ e
+      where
+        SomeException e = mkSomeExceptionWithLocation e
+
+Where ``mkSomeExceptionWithLocation`` (impurely) captures a backtrace and uses
+it to construct a ``SomeExceptionWithLocation``: ::
+
+    mkSomeExceptionWithLocation :: Exception e => e -> SomeException
+    mkSomeExceptionWithLocation e = unsafePerformIO $ do
+        bt <- collectBacktrace
+        return (SomeExceptionWithLocation bt e)
+
+Representing and capturing backtraces
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+GHC has at least three distinct mechanisms for capturing backtraces, each with
+their own backtrace representation:
+
+* ``HasCallStack`` is available in all programs, but requires modification of
+  the source program
+* the cost-centre profiler (accessible from Haskell with
+  ``GHC.Stack.CCS.getCurrentCCS``) can provide cost-center stacks
+* DWARF debug information in conjunction with GHC's `built-in stack unwinder
+  <https://www.haskell.org/ghc/blog/20200405-dwarf-3.html>`_ can be used
+  to provide terse (but nevertheless useful) backtraces with no runtime
+  overhead in the non-failing case (although backtrace collection is quite
+  slow)
+
+All of these backtrace options have their time and place, offering a range of
+levels of detail, executable size, and runtime overhead. GHC, being a compiler,
+shouldn't be in the business of dictating which of these mechanisms should be
+used to report exception backtraces.  Consequently, our ``Backtrace`` type is
+designed to capture them all: ::
+
+    -- | An exception backtrace.
+    data Backtrace
+        = CostCenterBacktrace (Ptr GHC.Stack.CCS.CostCentreStack)
+          -- ^ a cost center profiler backtrace
+        | HasCallStackBacktrace GHC.Stack.CallStack
+          -- ^ a stack from 'GHC.Stack.HasCallStack'
+        | ExecutionBacktrace [GHC.ExecutionStack.Location]
+          -- ^ a stack unwinding (e.g. DWARF) backtrace
+
+    instance Show Backtrace
+
+With such a type we can easily write a variant of ``throwIO`` that, for
+instance, attaches a DWARF backtrace: ::
+
+    -- | Throws an exception with a 'HasCallStack' backtrace.
```

Comment:
Isn't this a `HasCallStackBacktrace`, not a `ExecutionBacktrace` as implied above?  Also it's not clear how the Backtrace type enables this, at least not without a definition for `throw`.